src/profiler.cpp

```diff
@@ -1647,6 +1652,124 @@ void Profiler::dumpText(Writer& out, Arguments& args) {
+    _otlp_buffer.field(Otlp::ValueType::type_strindex, getOrSetIndex(string_idx_map, _engine->type()));
+    _otlp_buffer.field(Otlp::ValueType::unit_strindex, getOrSetIndex(string_idx_map, _engine->units()));
+    _otlp_buffer.field(Otlp::ValueType::aggregation_temporality, Otlp::AggregationTemporality::cumulative);
+    _otlp_buffer.commitMessage(sample_type_mark);
+
+    std::vector<CallTraceSample*> samples;
+    _call_trace_storage.collectSamples(samples);
+
+    u32 samples_counter_cache[samples.size()];
+    u32 samples_num_frames_cache[samples.size()];
+    size_t samples_idx = 0;
+
+    FrameName fn(args, args._style & ~STYLE_ANNOTATE, _epoch, _thread_names_lock, _thread_names);
+    protobuf_mark_t location_indices_mark = _otlp_buffer.startMessage(Otlp::Profile::location_indices);
+    for (auto it = samples.begin(); it != samples.end(); ++it) {
+        CallTrace* trace = (*it)->acquireTrace();
+        if (trace == NULL || excludeTrace(&fn, trace)) continue;
+
+        samples_counter_cache[samples_idx] = args._counter == COUNTER_SAMPLES ? (*it)->samples : (*it)->counter;
+        if (samples_counter_cache[samples_idx] == 0) continue;
+
+        samples_num_frames_cache[samples_idx] = trace->num_frames;
+        for (u64 j = 0; j < samples_num_frames_cache[samples_idx]; ++j) {
+            u64 function_idx = getOrSetIndex(function_idx_map, fn.name(trace->frames[j]));
+            _otlp_buffer.appendRepeated(function_idx);
+        }
+        ++samples_idx;
+    }
+    _otlp_buffer.commitMessage(location_indices_mark);
+
+    u64 frames_seen = 0;
+    for (samples_idx = 0; samples_idx < samples.size() && samples_counter_cache[samples_idx] > 0; samples_idx++) {
+        protobuf_mark_t sample_mark = _otlp_buffer.startMessage(Otlp::Profile::sample);
+        _otlp_buffer.field(Otlp::Sample::locations_start_index, frames_seen);
+        _otlp_buffer.field(Otlp::Sample::locations_length, samples_num_frames_cache[samples_idx]);
+        _otlp_buffer.field(Otlp::Sample::value, samples_counter_cache[samples_idx]);
+        _otlp_buffer.commitMessage(sample_mark);
+
+        frames_seen += samples_num_frames_cache[samples_idx];
+    }
+
+    _otlp_buffer.commitMessage(profile_mark);
+    _otlp_buffer.commitMessage(scope_profiles_mark);
+    _otlp_buffer.commitMessage(resource_profiles_mark);
+
+    protobuf_mark_t dictionary_mark = _otlp_buffer.startMessage(Otlp::ProfilesData::dictionary);
+
+    // Write mapping_table
+    // TODO: Include per-binary debug info
+    protobuf_mark_t mapping_mark = _otlp_buffer.startMessage(Otlp::ProfilesDictionary::mapping_table);
+    _otlp_buffer.commitMessage(mapping_mark);
+
+    // Convert function map (name to index) to an array (index to name mapping) for fast ordered access
+    const std::string* function_arr[function_idx_map.size()];
+    for (const auto& it : function_idx_map) {
+        function_arr[it.second] = &it.first;
+    }
```

Comment:
Can we have this logic in the `Index` class mentioned above?

Reply:
0ea647b581b0b7be0f0155459fa9bf62c08b741a